Update the managedBy KEP with fix for terminating pods (#4657)
kubernetes/enhancements: keps/sig-apps/4368-support-managed-by-for-batch-jobs/README.md

The API comments to the Job status API fields will be updated to make the contract
clear.

We also plan to add a validation rule to make sure that the terminal conditions
(`Failed` and `Complete`) are only added to Job when all pods are terminated.

> For that we plan to follow the approach described [below](#terminating-pods-and-terminal-job-conditions).

Author: Tentatively added to the validation seciton.
I think this seems like a useful extension of validation. It will fail fast in case some users want to write a controller which adds Failed or Complete directly, and it would prompt them to think about the termination of pods.

> 

Reviewer: Can we at least add 
- https://github.com/kubernetes/kubernetes/issues/125089
- https://github.com/kubernetes/kubernetes/issues/125185

as relevant bugs?

And mention to check them before the GA graduation? They do not need to necessarily block the graduation - just for visibility.

And can you also add any other relevant bugs that you might be aware of?

Author: I have mentioned them in the text for validation and in the graduation criteria. I think solving these issues blocks adding the validation rules actually.

> #### Terminating pods and terminal Job conditions

Author: I have added the points to the test plans. PTAL

During the development process of Alpha in 1.30 we considered adding a
validation rule enforcing that the Job terminal conditions (`Failed` or `Complete`)
are only added when all pods are terminated (`status.terminating=0` and
`status.ready=0`). However, the rule turned out to be violated by the built-in
Job controller (see issue [#123775](https://github.com/kubernetes/kubernetes/issues/123775)).

We are going to solve this issue by delaying the addition of the Job terminal
conditions (Failed or Complete) until the pods are terminated
(`terminating=0` and `ready=0`).

One complication of this approach is that pod termination may take an
arbitrarily long if the non-standard pod graceful termination period
(`terminationGracePeriodSeconds`) is configured (30s by default). In order to
give the API clients flexibility to know the fate of a Job as soon as possible
we extend the scope for the following Job conditions:
- `FailureTarget` - introduced in [Pod Failure Policy KEP](https://github.com/kubernetes/enhancements/issues/3329)

> - `SuccessCriteriaMet` - introduced in [Success Policy KEP](https://github.com/kubernetes/enhancements/issues/3998)

Reviewer 2: Again, I prefer links to KEPs themselves, rather than the issue. Less clicking through :wink:

Author: Sure, done

Note that, with the new conditions an API client can know the fate of a Job
earlier than currently, since adding the conditions will not wait for emptying
of the `uncountedTerminatedPods` structure, as shown in the
[experiment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2114710754).

Since the fix is needed for this KEP and the
[Pod Replacement Policy KEP](https://github.com/kubernetes/enhancements/issues/3939),
we intend to protect the fix with the OR of the `JobManagedBy` and the
`JobReplacementPolicy` feature gates.

> Note also that the fix impacts `CronJob` which in the `Forbid` concurrency

Reviewer 2: ```suggestion
Note also that the fix impacts `CronJob` when using the `Forbid` concurrency
```

policy (see [here](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#concurrency-policy)).
However, it can be argued that the current behavior of the CronJob / Job integration
does not match the expectations of the `Forbid` policy
(see [comment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2115200217)),

> and can be handled as a bug.

Author: I add this note tentatively to non-goals.

I would be ok to support the new concurrency policy `ForbidActive` (or `SoftForbid`), but it isn't clear if needed, since the proposed changes better much the current documentation for the `Forbid` mode. OTOH, it might be a useful bailout config option in case there is a user who want to have the semantic as in your point 1.

cc @deads2k @soltysh @alculquicondor in case you have some additional opinion.

> and can be handled as a bug.

Reviewer 3: I remember this from the success policy KEP.  As I recall, we chose to make the SuccessPolicy indicate `SuccessCriteriaMet` when we met criteria, but wait until all pods were terminated before transitioning to `Complete`.  Since the CronJob waits for complete, this makes the cronjob controller wait. Additionally, this is the safe default.

I'm ok with the idea of an addition concurrency policy, but I see it as separate from this KEP.

> and can be handled as a bug.

Author: > Maybe it would be good to also mention the pods with long termination times to the Risks section?

Done. PTAL.

The validation rule will be added in the +1 release (1.32) to ensure the
addition of the `Failed` and `Complete` conditions is delayed appropriately in
case of a skewed deployment when the kube-controller-manager is one version

> ahead the kube-apiserver.

Reviewer 2: Due to this fact, wouldn't it be better to stick with alpha, which is off by default and only turn it on when this validation is in place? This will save potential problems. Alternatively, we can promote this to Beta, but keep the feature off by default. Either is fine, as long as we don't enable this before adding this validation.

Author: I think second alpha would be the preference as the feature gate off and beta isn't typical. I update the relevant places:
- the prod readiness yaml
- kep.yaml
- graduation steps

Reviewer 4: It is still within bounds to change behavior while in Beta. Also, this will just affect external job controllers that probably don't exist yet. We can leave some notes in the documentation about the validation that we plan to add, for anyone that wishes to implement a job controller using managedBy.

Keeping the feature in alpha (or beta disabled by default) will delay the adoption of MultiKueue with Jobs for additional 4 months.

Reviewer 5: I would say that "beta off-by-default" doesn't give us anything and isn't what I would seriously consider.

That said, we are already stating clearly that we will harden the validation in the next release. People are generally not reading documentation and given that I think we should stick to second Alpha - we shouldn't consciously go for a choice where we know we will break some potential users of Beta API - that just doesn't sound right to me.

Reviewer 3: Agree with @wojtek-t.  Since this is remaining off-by-default (your explanation is very good), I would stay in Alpha.

[experiment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2114710754).

Since the fix is needed for this KEP and the
[Pod Replacement Policy KEP](https://github.com/kubernetes/enhancements/issues/3939),
we intend to protect the fix with the OR of the `JobManagedBy` and the
`JobReplacementPolicy` feature gates.

Additionally, fixing the issue [#123775](https://github.com/kubernetes/kubernetes/issues/123775) will
also require fixing:
- [Job controller reports the count of terminating pods with unnecessary delay](https://github.com/kubernetes/kubernetes/issues/125089) and
- [Job controller reports the count of ready pods with unnecessary delay](https://github.com/kubernetes/kubernetes/issues/125185).

Note also that the fix impacts `CronJob` when using the `Forbid` concurrency
policy (see [here](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#concurrency-policy)).
However, it can be argued that the current behavior of the CronJob / Job integration
does not match the expectations of the `Forbid` policy
(see [comment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2115200217)),
and can be handled as a bug.

> The validation rule will be added in the +1 release (1.32) to ensure the

Reviewer 3: Since we're remaining in Alpha, it seems acceptable to me that we add the validation rule now and accept the hiccup during upgrades with the alpha enabled.  Doing this will allow those developing new job controllers to see the correct behavior in 1.31 instead of waiting until 1.32.  The cost of doing this is a few fail update/status during rolling updates of alpha enabled clusters.

Reviewer 5: +1

Author: I see, it makes perfect sense. Initially I added validation to the graduation criteria for Beta, cause while nice-to-have it does not seem required for alpha.

After the comment I moved it to the second alpha section to better reflect the intention of working on it during the second alpha.

- e2e tests
- implement the `job_by_external_controller_total` metric
- address all known inconsistencies between validation and the Job controller behavior,
  in particular: [#123775](https://github.com/kubernetes/kubernetes/issues/123775).
  The proposed approach is outlined in [here](#terminating-pods-and-terminal-job-conditions).
- verify the validation passes during e2e tests for open-source projects (like Kueue and JobSet)
- The feature flag enabled by default

Second Beta (1.32):
- Add validation rule that `Failed` and `Complete` conditions are added when

>   `terminating=0`, and `ready=0`.

Reviewer 2: Either second beta, where we'd turn this on, first beta would be off by default to ensure the validation is in place. Or second alpha, and beta only in 1.31. See my other comment about it.

Author: As in the other comment, I think a second alpha would me more intuitive to users. Updated.

- the `job_by_external_controller_total` metric is not incremented for a new Job without "managedBy" or with default value
- the `job_by_external_controller_total` metric is not incremented for Job updates (regardless of the "managedBy")

The following scenarios related to [Terminating pods and terminal Job conditions](#terminating-pods-and-terminal-job-conditions) are covered:
- `Failed` or `Complete` conditions are not added while there are still terminating pods
- `FailureTarget` is added when backoffLimitCount is exceeded, or activeDeadlineSeconds timeout is exceeded

> - `SuccessCriteriaMet` is added when the `completions` are satisfied

Reviewer 6: I have a question about responsibility. This change is needed to graduate the SuccessPolicy (KEP3998)? Or we can graduate the SuccesPolicy regardless of this change?

Author: IMO we can graduate independently. The new fix reuses the already introduced conditions which match the purpose.

Reviewer 6: That makes sense. 
The `SuccessCriteriaMet` condition is still guarded by the SuccessPolicy feature gate, right?
I meant the new transition is not guarded by the ManagedBy feature gate, right?

Author: Depends on the context. When adding it for meeting success policy we guard it by the success policy feature gate. When adding for meeting completions we guard it by managedBy or podreplacementpplicy feature gates. The code which translates SuccessCriteriaMet into Complete condition will be guarded by OR of the feature gates.

Reviewer 6: That makes sense. Maybe we can revisit here once we graduate the successPolicy to GA.

- 2023-12.20 - First version of the KEP
- 2024-03-05 - Merged implementation PR [Support for the Job managedBy field (alpha)](https://github.com/kubernetes/kubernetes/pull/123273)

> - 2024-03-07 - Merged [Update Job conformance test for job status updates](https://github.com/kubernetes/kubernetes/pull/123751)

Reviewer 5: non-blocking PRR things:

> Are there any tests for feature enablement/disablement?

Were the tests added? If not, can we prioritize them for the second Alpha?

Author: Yes, we have the unit tests for enablement of the feature: https://github.com/kubernetes/kubernetes/blob/169a952720ebd75fcbcb4f3f5cc64e82fdd3ec45/pkg/controller/job/job_controller_test.go#L2365

We also have an integration test for re-enabling the feature: https://github.com/kubernetes/kubernetes/blob/169a952720ebd75fcbcb4f3f5cc64e82fdd3ec45/test/integration/job/job_test.go#L1691.

Reviewer 5: > Yes, we have the unit tests for enablement of the feature: https://github.com/kubernetes/kubernetes/blob/169a952720ebd75fcbcb4f3f5cc64e82fdd3ec45/pkg/controller/job/job_controller_test.go#L2365

These aren't enablement tests - these are feature tests.

Please take a look at KEP template - I'm looking for simple tests that change the value of FG in the middle of the tests.

Author: > Please take a look at KEP template - I'm looking for simple tests that change the value of FG in the middle of the tests.

I think the integration test satisfies this expectation.  See the flips in the middle [disabling](https://github.com/kubernetes/kubernetes/blob/169a952720ebd75fcbcb4f3f5cc64e82fdd3ec45/test/integration/job/job_test.go#L1745), [re-enabling](https://github.com/kubernetes/kubernetes/blob/169a952720ebd75fcbcb4f3f5cc64e82fdd3ec45/test/integration/job/job_test.go#L1763)

I don't think we have a unit test like that, do you think we should also add unit, given we have integration?

Reviewer 5: missed the integration, sorry. This is fine.

Just please link it here in this question explicitly.

Author: Sure, updated with the link
